[export][serde] Add opset version check and upgrader API

larryliu0820 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #103238


This PR adds initial implementation of an upgrader. Added test to show that this version works for one of the upgraders in https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/operator_upgraders/upgraders_entry.cpp.
    


Differential Revision: [D46651778](https://our.internmc.facebook.com/intern/diff/D46651778)

larryliu0820 committed:
[RFC] op upgrader API

[ghstack-poisoned]

larryliu0820 committed:
Update on "[RFC] op upgrader API"




[ghstack-poisoned]

larryliu0820 commented on torch/_export/serde/serialize.py:
```diff
@@ -633,6 +635,76 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+        compiler_opset_version: Optional[Dict[str, int]] = None,
+        model_opset_version: Optional[Dict[str, int]] = None,
+    ):
+        op_upgraders = {}  # can add a new TS API: torch._C._get_upgraders_entry_map()
```
Yeah this is referring to torch script upgraders. Current approach is ask the developer to add upgraders to TorchScript the same way as before, they can also be applied to Export without extra work.

larryliu0820 commented on torch/_export/serde/serialize.py:
```diff
@@ -873,6 +945,10 @@ def deserialize(
             .deserialize(serialized_exported_program.graph_module)
         )
+        model_opset_version: Dict[str, int] = {}  # TODO(larryliu) this should be deserialized as well
+        upgrader = GraphModuleOpUpgrader(self.expected_opset_version, model_opset_version)
```
It needs model op set version which is only available here.

angelayi replied:
Hmm ok...btw model opset version is located in `serialized_exported_program.opset_version`

avikchaudhuri commented:
Can you set up a way to write tests for this infra? E.g., have an end-to-end test where an operator goes through a sequence of versions, and a simple model goes through an upgrade process a couple of times.

larryliu0820 committed:
Update on "[RFC] op upgrader API"




[ghstack-poisoned]

avikchaudhuri commented:
@avikchaudhuri has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46651778).

avikchaudhuri commented:
@larryliu0820 I imported this diff to play with it and I'm getting the same error as the failing tests:
```
torch/_export/serde/serialize.py", line 686, in _parse_upgraders
    model_ver = self.model_opset_version["aten"]
KeyError: 'aten'
```

What's the fix?

larryliu0820 committed:
Update on "[RFC] op upgrader API"


Differential Revision: [D46651778](https://our.internmc.facebook.com/intern/diff/D46651778)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[RFC] op upgrader API"


Differential Revision: [D46651778](https://our.internmc.facebook.com/intern/diff/D46651778)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[RFC] op upgrader API"


Differential Revision: [D46651778](https://our.internmc.facebook.com/intern/diff/D46651778)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[export][serde] Add opset version check and upgrader API"


This PR adds two things:

1. a validator to reconcile model opset version and deserializer opset version. This currently fails if any of the version doesn't match.
2. initial implementation of an upgrader. Added test to show that this version works for one of the upgraders in https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/operator_upgraders/upgraders_entry.cpp.

Note: the validator will only print warning if the namespace in model is missing from the deserializer.



Differential Revision: [D46651778](https://our.internmc.facebook.com/intern/diff/D46651778)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[export][serde] Add opset version check and upgrader API"


This PR adds two things:

1. a validator to reconcile model opset version and deserializer opset version. This currently fails if any of the version doesn't match.
2. initial implementation of an upgrader. Added test to show that this version works for one of the upgraders in https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/operator_upgraders/upgraders_entry.cpp.

Note: the validator will only print warning if the namespace in model is missing from the deserializer.



Differential Revision: [D46651778](https://our.internmc.facebook.com/intern/diff/D46651778)

[ghstack-poisoned]

zhxchen17 commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,150 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+            try:
+                register_old_op(name=upgrader_name, schema=schema, impl_str=upgrader_str)
+            except RuntimeError as e:
+                if "with the same name and overload name multiple times" in str(e):
```
lol is it possible to have a special error type so we don't have match error message specifically?

larryliu0820 replied:
yeah this is part of torch.library.Library API. I can update it in later PRs.

zhxchen17 commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,150 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+        self.compiler_opset_version = compiler_opset_version if compiler_opset_version else {}
+        self.model_opset_version = model_opset_version if model_opset_version else {}
+        # key is version number, value is a list of upgraders, keyed on op name
+        self.upgraders: List[Tuple[str, str]] = self._parse_upgraders(op_upgraders)
```
does this need to be an attribute? We can directly pass this to _populate_passes()

zhxchen17 reviewed (commented):
I feel upgrader kinda deserves its own python module separately since there isn't much coupling between upgrader and main deserialization logic.

avikchaudhuri commented on test/export/test_serialize.py:
```diff
@@ -7,12 +7,21 @@
+  if (self.is_floating_point() or isinstance(other, float)):
+    return self.true_divide(other)
+  return self.divide(other, rounding_mode='trunc')
+        """)
```
format nit: 
* comma after triple quote
* close paren on next line

avikchaudhuri commented on test/export/test_serialize.py:
```diff
@@ -199,6 +208,52 @@ def forward(self, x):
+
+        def count_op(graph, target_str):
+            return len([n for n in graph.nodes if isinstance(n.target, torch._ops.OpOverload) and n.target.name() == target_str])
+        count = count_op(ep.graph, "aten::div.Scalar")
```
nice, maybe just count before the `ep.transform(...)` and assert `== 1`?

avikchaudhuri commented on test/export/test_serialize.py:
```diff
@@ -199,6 +208,52 @@ def forward(self, x):
         inputs = (torch.ones([512], requires_grad=True),)
         self.check_graph(MyModule(), inputs)
 
+class TestOpVersioning(TestCase):
```
I don't see a test for version check, did I miss something?

larryliu0820 replied:
test_empty_model_opset_version_raises test_opset_mismatch_raises test_model_op_namespace_version_missing_from_deserializer_do_not_raises

avikchaudhuri replied:
Sorry, where are these tests? Are they existing or in this PR, and if the latter, I can't see them in the changed files lol. Did you forget to add them to the commit?

larryliu0820 replied:
Sorry I just split this PR into 2. See the base PR for version mismatch tests

avikchaudhuri replied:
Right, of course. Duh. :D

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+  if (self.is_floating_point() or isinstance(other, float)):
+    return self.true_divide_(other)
+  return self.divide_(other, rounding_mode='trunc')
+",
```
nit: add some tabs here to match the opening quote

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+            self,
+            compiler_opset_version: Optional[Dict[str, int]] = None,
+            model_opset_version: Optional[Dict[str, int]] = None,
+            op_upgraders: Optional[Dict[str, Tuple[str, ...]]] = None,
```
Not sure I follow the type here: what's contained in the Tuple here? 
* first element is signature?
* second element is code?
* and then...what?

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+            self.new_target = new_target
+
+        def __call__(self, fx_module: GraphModule) -> Optional[PassResult]:
+            replace_target_nodes_with(fx_module, "call_function", self.old_target, "call_function", self.new_target)
```
for my benefit can you describe in a comment what `replace_target_nodes_with`, or this particular call of it anyway, does?

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+
+        self.upgrader_passes: List[GraphModuleOpUpgrader.UpgraderPass] = self._populate_passes()
+
+    def _parse_upgraders(self, op_upgraders: Dict[str, Tuple[str]] = None) -> List[Tuple[str]]:
```
Tuple type inside `op_upgraders` doesn't match that in the constructor

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+        """reorder op_upgraders by version number."""
+        if not op_upgraders:
+            return []
+        model_ver = self.model_opset_version["aten"]
```
what's special about `aten` here? If it's special, shouldn't we insist (i.e., early-validate) that opset versions always have `aten`?

larryliu0820 replied:
ATen op set versioning is something we had for TorchScript and I'm making sure it works in Export. We will add custom opset versioning in next PRs. But good point on validating it.

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+
+        def get_target_version(versioned_upgrader_name: str) -> int:
+            """div_Scalar_0_3 is the name of the upgrader, meaning it applies to opset version 0 to 3 and is
+            upgrading to version 4."""
```
> upgrading to version 4

* what about version `6`? do you need a `0_5` or a `4_5` (to chain with `0_3`)?

larryliu0820 replied:
> upgrading to version 4

This is just saying bumping from 3 to 4 because of a schema change in `div.Scalar`. If we soon have another update to `div.Scalar` then the name will be `div__Scalar_4_4`, meaning it can only be applied to opset version 4. If we instead update another op say `mul.Scalar`, the upgrader can be `mul__Scalar_0_4`.

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+        model_ver = self.model_opset_version["aten"]
+        curr_ver = self.compiler_opset_version["aten"]
+
+        def get_target_version(versioned_upgrader_name: str) -> int:
```
missing various validation checks: e.g., for `X_Y`, must `X` < `Y` or must `X` <= `Y` (i.e., can I have `X_X` or should I just write `X` as a suffix)?

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+
+        versioned_upgraders: Dict[int, Tuple[str]] = {get_target_version(name): v for name, v in op_upgraders.items()}
+        target_upgraders: List[Tuple[str]] = []
+        for ver in range(model_ver, curr_ver + 1):
```
This logic is subtle and seems to be doing some general thing where multiple upgraders are provided. Please add some test that brings out the subtlety here.

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+
+    def _populate_passes(self) -> List[UpgraderPass]:
+        """Given a list of upgraders, loop through it from lower version to higher version and create passes for all
+        upgraders. se torch.Library API to register old ops. Op name will be
```
se -> Use?

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+    def _populate_passes(self) -> List[UpgraderPass]:
+        """Given a list of upgraders, loop through it from lower version to higher version and create passes for all
+        upgraders. se torch.Library API to register old ops. Op name will be
+        <name>_<valid_from_ver>_<valid_till_ver>. Register upgarders as CompositeImplicitAutograd kernels. For example:
```
till -> to?
Also please comment on whether the (from, to) range is inclusive or exclusive

avikchaudhuri replied:
typo: upgarders

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+            impl_lib.impl(name, locals()[name], "CompositeImplicitAutograd")
+
+        for schema, upgrader_str in self.upgraders:
+            upgrader_name = upgrader_str.split('(')[0].split(' ')[-1]
```
pull this kind of parsing logic into a util and add tests for it, please

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+            try:
+                register_old_op(name=upgrader_name, schema=schema, impl_str=upgrader_str)
+            except RuntimeError as e:
+                if "with the same name and overload name multiple times" in str(e):
```
hmm, this is pretty flaky

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+                else:
+                    raise RuntimeError from e
+            old_op_target = getattr(torch.ops.aten, upgrader_name).default
+            op_name, overload_name = (op_name, "default") if "." not in op_name else tuple(op_name.split(".")[:2])
```
comment with example, please

avikchaudhuri commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+        from torch._dynamo import config
+        config.traceable_tensor_subclasses.add(FakeTensor)
+        for _pass in self.upgrader_passes:
+            exported_program = exported_program.transform(_pass).transform(partial(export, args=inputs))
```
What's the second transform again? Export again? If so, shouldn't it happen once at the very end? Also missing other things passed to export, such as `constraints`?

Unless I completely misunderstood...

larryliu0820 replied:
Added comments, hopefully it's more clear.

avikchaudhuri reviewed (changes requested):
This is a great start! Requesting changes mainly for more tests and docs, see inline comments.

larryliu0820 commented on test/export/test_serialize.py:
```diff
@@ -199,6 +208,52 @@ def forward(self, x):
+        count = count_op(ep.graph, "aten::div.Scalar")
+        self.assertEqual(count, 1)
+
+        upgraded_ep = upgrader.upgrade(ep)
```
So the upgrader have 2 steps: 1. replace op with old version op, 2. retrace. After step 1, we should have the old version op (`div__Scalar_mode_0_3` in this example). After step 2 this op should be decomposed into the ops in the upgrader implementation.

larryliu0820 commented on test/export/test_serialize.py:
```diff
@@ -199,6 +208,52 @@ def forward(self, x):
+        upgraded_ep = upgrader.upgrade(ep)
+
+        custom_op_count = count_op(upgraded_ep.graph, "aten::div__Scalar_0_3")
+        self.assertEqual(custom_op_count, 0)
```
See comment above

larryliu0820 commented on torch/_export/serde/serialize.py:
```diff
@@ -633,15 +632,150 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+            upgraded_program = exported_program.transform(_pass)
+            # NB: we have to retrace the graph_module instead of ep because of some failure.
+            exported_program = export(upgraded_program.graph_module, inputs, [])
+            exported_program.call_spec = upgraded_program.call_spec
```
Doesn't work for some reason

angelayi replied:
oh right...sorry 😭

larryliu0820 committed:
Update on "[export][serde] Add opset version check and upgrader API"


This PR adds two things:

1. a validator to compare model opset version and deserializer opset version. This currently raises exception if any of the version doesn't match.
2. initial implementation of an upgrader. Added test to show that this version works for one of the upgraders in https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/operator_upgraders/upgraders_entry.cpp.

Note: the validator will only print warning if the namespace in model is missing from the deserializer.



Differential Revision: [D46651778](https://our.internmc.facebook.com/intern/diff/D46651778)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[export][serde] Add opset version check and upgrader API"



This PR adds initial implementation of an upgrader. Added test to show that this version works for one of the upgraders in https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/operator_upgraders/upgraders_entry.cpp.
    


Differential Revision: [D46651778](https://our.internmc.facebook.com/intern/diff/D46651778)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[export][serde] Add opset version check and upgrader API"



This PR adds initial implementation of an upgrader. Added test to show that this version works for one of the upgraders in https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/operator_upgraders/upgraders_entry.cpp.
    


Differential Revision: [D46651778](https://our.internmc.facebook.com/intern/diff/D46651778)

[ghstack-poisoned]

larryliu0820 commented on test/export/test_serialize.py:
```diff
@@ -199,6 +208,52 @@ def forward(self, x):
+
+        compiler_opset_version = {"aten": 4}
+        model_opset_version = {"aten": 3}
+        upgrader = GraphModuleOpUpgrader(compiler_opset_version, model_opset_version, TEST_UPGRADERS)
```
Added tests

avikchaudhuri commented on torch/_export/serde/upgrade.py:
```diff
@@ -0,0 +1,175 @@
+        upgrading function itself). After retrace, this custom operator will be decomposed into the ops used in the
+        upgrader. After all passes are applied, the exported program will be upgraded to the target version."""
+        args = [n.meta["val"] for n in exported_program.graph.nodes if n.op == "placeholder"]
+        args_real_tensors = [torch.ones(tuple(arg.size()), dtype=arg.dtype) if isinstance(arg, FakeTensor) else arg for
```
This line is causing tests to fail. Do we have a reference to original args? @angelayi

larryliu0820 replied:
Yeah it will be great if we can serialize the original input.

avikchaudhuri commented:
Accepting to unblock, happy with the changes since last review.

larryliu0820 committed:
Update on "[export][serde] Add opset version check and upgrader API"



This PR adds initial implementation of an upgrader. Added test to show that this version works for one of the upgraders in https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/operator_upgraders/upgraders_entry.cpp.
    


Differential Revision: [D46651778](https://our.internmc.facebook.com/intern/diff/D46651778)

[ghstack-poisoned]

larryliu0820 committed:
Update on "[export][serde] Add opset version check and upgrader API"



This PR adds initial implementation of an upgrader. Added test to show that this version works for one of the upgraders in https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/operator_upgraders/upgraders_entry.cpp.
    


Differential Revision: [D46651778](https://our.internmc.facebook.com/intern/diff/D46651778)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46651778).

larryliu0820 commented:
@pytorchbot merge

larryliu0820 commented:
@pytorchbot merge
